tasm-lib/src/verifier/fri/verify_fri_authentication_paths.rs

```diff
@@ -364,48 +360,94 @@ mod test {
-        }
+    #[proptest]
+    fn fri_authentication_fails_if_xor_bitflag_is_disturbed_slightly(seed: [u8; 32]) {
+        let mut initial_state = VerifyFriAuthenticationPaths.pseudorandom_initial_state(seed, None);
+        let top_of_stack = initial_state.stack.len() - 1;
+        let xor_bitflag = initial_state.stack.get_mut(top_of_stack - 8).unwrap();
+        *xor_bitflag *= bfe!(2); // todo: generalize this perturbation
+        prop_assume!(u32::try_from(*xor_bitflag).is_ok());
+
+        test_assertion_failure(
+            &ShadowedAlgorithm::new(VerifyFriAuthenticationPaths),
+            initial_state.into(),
+            &[30],
+        );
+    }
+
+    #[proptest]
+    fn fri_authentication_fails_if_authentication_path_is_disturbed_slightly(
+        seed: [u8; 32],
+        digest_index: usize,
+        #[strategy(0_usize..5)] perturbation_index: usize,
+        #[filter(#perturbation != 0)] perturbation: i8,
+    ) {
+        let mut initial_state = VerifyFriAuthenticationPaths.pseudorandom_initial_state(seed, None);
+        let auth_paths = &mut initial_state.nondeterminism.digests;
+        let digest_index = digest_index % auth_paths.len();
+        let auth_path_element = &mut auth_paths[digest_index];
+        let Digest(ref mut auth_path_element_innards) = auth_path_element;
+        auth_path_element_innards[perturbation_index] += bfe!(perturbation);
+
+        test_assertion_failure(
+            &ShadowedAlgorithm::new(VerifyFriAuthenticationPaths),
+            initial_state.into(),
+            &[30],
+        );
+    }
+
+    #[proptest]
+    fn fri_authentication_fails_if_a_index_is_disturbed_slightly(
+        seed: [u8; 32],
+        perturbation_index: usize,
+        #[filter(#perturbation != 0)] perturbation: i8,
+    ) {
+        let perturbation = bfe!(perturbation);
+
+        let mut initial_state = VerifyFriAuthenticationPaths.pseudorandom_initial_state(seed, None);
+        let top_of_stack = initial_state.stack.len() - 1;
+
+        let a_indices_pointer = initial_state.stack[top_of_stack - 6];
+        let a_indices_len = initial_state.nondeterminism.ram[&a_indices_pointer].value() as usize;
+        let perturbation_index = bfe!(perturbation_index % a_indices_len);
+
+        let perturbation_pointer = a_indices_pointer + bfe!(1) + perturbation_index;
+        let ram = &mut initial_state.nondeterminism.ram;
+        let a_index = ram.get_mut(&perturbation_pointer).unwrap();
+
+        let old_a_index = a_index.value();
+        *a_index += perturbation;
+        let new_a_index = a_index.value();
+        prop_assume!(u32::try_from(*a_index).is_ok());
+
+        // ensure meaningful perturbation
+        let dom_len_minus_one = initial_state.stack[top_of_stack - 9].value();
+        let xor_bitflag = initial_state.stack[top_of_stack - 8].value();
+        let node_index = |i| (i & dom_len_minus_one) ^ xor_bitflag;
+        prop_assume!(node_index(old_a_index) != node_index(new_a_index));
```

Comment:
@Sword-Smith This is required for the proptest to work. I'm unsure whether it indicates a bug in the snippet: should out-of-range leaf indices be accepted?

Reply:
Right. These arguments are used for some fancy bit-magic to make the index calculations more efficient in FRI? 

The leaf indices are mapped onto the correct range with a bit-flag, `dom_len_minus_one`. So `i` (or `a_index`) is definitely allowed to be outside the range of indices for this round. And since all indices are derived from a sampling function that maps value to the correct range, I would say this code is correct.